### PR TITLE
ci: bring up benchmark

### PR DESCRIPTION
**Pull request by SukkaW**
## What does it do?

Add a `benchmark.sh` under `test` directory, used to perform benchmark for every build on travis ci, which enables us to evaluate the generation speed after each commit.

## How to test

```sh
git clone -b benchmark https://github.com/sukkaw/hexo.git
cd hexo
npm install
./test/benchmark.sh
```

## Screenshots

image

## Pull request tasks

- [ ] Add test cases for the changes.
- [x] Passed the CI test.

### CODE REVIEW

File: test/benchmark.sh [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+}
+
+LOG_TABLE () {
+    time_begin=$(date +%s -d "$(awk '/.*DEBUG Hexo version/{print $1}' build.log)")
+    time_process_start=$(date +%s.%3N -d "$(awk '/.*INFO  Start processing/{print $1}' build.log)")
+    time_render_start=$(date +%s.%3N -d "$(awk '/.*INFO  Files loaded in/{print $1}' build.log)")
+    time_render_finish=$(date +%s.%3N -d "$(awk '/.*INFO.*generated in /{print $1}' build.log)")
+    time_database_saved=$(date +%s.%3N -d "$(awk '/.*DEBUG Database saved/{print $1}' build.log)")
+
+    echo "Load Plugin/Scripts/Database $(_SUBSTRUCTION $time_process_start $time_begin)\nProcess Source $(_SUBSTRUCTION $time_render_start $time_process_start)\nRender Files $(_SUBSTRUCTION $time_render_finish $time_render_start)\nSave Database $(_SUBSTRUCTION $time_database_saved $time_render_finish)" | awk '{printf "| %-28s | %7.3fs |\n",$1" "$2,$3"s"}'
```
Comment (YoshinoriN) [inline test/benchmark.sh]: IMHO this line too long. How about split like following?

```sh
_MESSAGE_FORMATTER () {
    awk '{printf "| %-28s | %7.3fs |\n",$1" "$2,$3"s"}'
}

LOG_TABLE () {
    time_begin=$(date +%s -d "$(awk '/.*DEBUG Hexo version/{print $1}' build.log)")

    ...

    load_message="Load Plugin/Scripts/Database $(_SUBSTRUCTION $time_process_start $time_begin)"
    process_source_message="Process Source $(_SUBSTRUCTION $time_render_start $time_process_start)"
    render_files_message="Render Files $(_SUBSTRUCTION $time_render_finish $time_render_start)"
    save_database_message="Save Database $(_SUBSTRUCTION $time_database_saved $time_render_finish)"

    echo $load_message | _MESSAGE_FORMATTER
    echo $process_source_message | _MESSAGE_FORMATTER
    echo $render_files_message | _MESSAGE_FORMATTER
    echo $save_database_message | _MESSAGE_FORMATTER
}
```

Comment (SukkaW) [reply test/benchmark.sh]: Updated.

**Comment by SukkaW** [comment]
> How about add `Total time`?
> 
> ```shell
> "Total time $(_SUBSTRUCTION $time_database_saved $time_begin)"
> ```

Added.

**Comment by SukkaW** [comment]
@curbengh I'd like to add `time` in the `benchmark.sh`, but it is a little tricky to redirect `time` output to a file. I will have a try.

**Comment by SukkaW** [comment]
The memory usage is now added

File: .travis.yml [hunk 2]
```diff
@@ -2,6 +2,11 @@ language: node_js
 
 sudo: false
 
+addons:
+  apt:
+    packages:
+    - time
```
Comment (curbengh) [inline .travis.yml]: I thought this is built-in? I'm [able to run](https://travis-ci.com/curbengh/hexo-site/jobs/247537552) `/usr/bin/time` without adding this package. I'm using bionic in that test, perhaps it's not included in xenial?

File: test/benchmark.sh [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+    time_database_saved=$(date +%s.%3N -d "$(awk '/.*DEBUG Database saved/{print $1}' build.log)")
+
+    memory_usage=$(awk '/.*Maximum resident set size/{print $6}' build.log)
+
+    echo "Load Plugin/Scripts/Database $(_SUBSTRUCTION $time_process_start $time_begin)s" | _MESSAGE_FORMATTER
+    echo "Process Source $(_SUBSTRUCTION $time_render_start $time_process_start)s" | _MESSAGE_FORMATTER
+    echo "Render Files $(_SUBSTRUCTION $time_render_finish $time_render_start)s" | _MESSAGE_FORMATTER
+    echo "Save Database $(_SUBSTRUCTION $time_database_saved $time_render_finish)s" | _MESSAGE_FORMATTER
+    echo "Total time $(_SUBSTRUCTION $time_database_saved $time_begin)s" | _MESSAGE_FORMATTER
+    echo "Memory Usage(RSS) $(echo | awk "{printf \"%.3f\", $memory_usage/1024}")MB" | _MESSAGE_FORMATTER
```
Comment (curbengh) [inline test/benchmark.sh]: 🤔 what is RSS in this context?

File: .travis.yml (hunk 2 above)
Comment (YoshinoriN) [reply .travis.yml]: I had the same question. And I investigated. The ubuntu 16.04 server seems not installed `time` package by default.

> http://releases.ubuntu.com/16.04/
> http://releases.ubuntu.com/16.04/ubuntu-16.04.6-server-amd64.manifest

File: test/benchmark.sh (hunk 3 above)
Comment (SukkaW) [reply test/benchmark.sh]: `Resident Set Size`

Comment (YoshinoriN) [reply test/benchmark.sh]: > [What is RSS and VSZ in Linux memory management](https://stackoverflow.com/questions/7880784/what-is-rss-and-vsz-in-linux-memory-management)

**Pull request merged**